modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2ManagedLocalResult.java

```diff
@@ -43,71 +86,546 @@
+            external.reset();
+        }
+    }
 
-        if (memory < 0)
-            mem.release(-memory);
-        else
-            mem.reserve(memory);
+    /** {@inheritDoc} */
+    @Override public Value[] currentRow() {
+        return currentRow;
+    }
 
-        memReserved += memory;
+    /** {@inheritDoc} */
+    @Override public boolean next() {
+        if (!closed && rowId < rowCount) {
+            rowId++;
+            if (rowId < rowCount) {
+                if (external != null) {
+                    currentRow = external.next();
+                } else {
+                    currentRow = rows.get(rowId);
+                }
+                return true;
+            }
+            currentRow = null;
+        }
+        return false;
+    }
+
+    /** {@inheritDoc} */
+    @Override public int getRowId() {
+        return rowId;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isAfterLast() {
+        return rowId >= rowCount;
+    }
+
+    /**
+     * @param values Values.
+     */
+    private void cloneLobs(Value[] values) {
+        for (int i = 0; i < values.length; i++) {
+            Value v = values[i];
+            Value v2 = v.copyToResult();
+            if (v2 != v) {
+                containsLobs = true;
+                session.addTemporaryLob(v2);
+                values[i] = v2;
+            }
+        }
+    }
+
+    /**
+     * @param values row.
+     * @return Row,
+     */
+    private ValueRow getDistinctRow(Value[] values) {
+        if (distinctIndexes != null) {
+            int cnt = distinctIndexes.length;
+            Value[] newValues = new Value[cnt];
+            for (int i = 0; i < cnt; i++) {
+                newValues[i] = values[distinctIndexes[i]];
+            }
+            values = newValues;
+        }
+        else if (values.length > visibleColumnCount) {
+            values = Arrays.copyOf(values, visibleColumnCount);
+        }
+        return ValueRow.get(values);
+    }
+
+    private void createExternalResult() {
+        assert ctx != null;
+
+        external = distinct || distinctIndexes != null || sort != null ?
+            new SortedExternalResult(ctx, session, distinct, distinctIndexes, visibleColumnCount, sort, memTracker,
+                rowCount) : new PlainExternalResult(ctx, memTracker);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void addRow(Value[] values) {
+        cloneLobs(values);
+        if (isAnyDistinct()) {
+            if (distinctRows != null) {
+                ValueRow array = getDistinctRow(values);
+                Value[] previous = distinctRows.get(array);
+                if (previous == null || sort != null && sort.compare(previous, values) > 0) {
+                    distinctRows.put(array, values);
+                }
+                rowCount = distinctRows.size();
+                if (!isEnoughMemoryAfterReservation(array, previous, values)) {
+                    createExternalResult();
```

Comment:
I suppose the code is borrowed from h2 but I suspect that external result could be recreated here if one already exists. I would good to ensure somehow that there is no mistake. Simple thing coming to mind is `assert external == null`.

Reply:
Actually it cannot be recreated because there is of invariant `distinctRows != null XOR external != null`. So only either of them is not null.